### DISCUSSION TITLE
Writing Rust Modules

### DISCUSSION THREAD

**Discussion by thiskappaisgrey** [emacs-ng/emacs-ng#477] [Q&A] [answered]
Hi, 
How do I write / compile Rust modules for my Emacs  config? I kind of want to do some experiments with Emacs structural editing using the Rust Tree-Sitter libraries (since tree-sitter in Emacs is natively supported) eventually. Is it possible to write / compile rust modules in a separate package like you can with C? I'm not really familiar with the Emacs dynamic FFI yet, so I'm not really sure where to look in the code base for examples of this. Tysm!

**Comment by DavidDeSimone** [comment, marked as answer]
Hello - a couple of follow up questions. Do you want to compile Rust modules for mainline emacs, or do you want to make a Rust module in emacs-ng? 

If you want to make a standalone Rust module that can be integrated into mainline emacs, there is the excellent repo https://github.com/ubolonton/emacs-module-rs which contains a Rust package that will allow you to produce a compatible shared library that can be used in mainline emacs. I have a personal example of such a module: https://github.com/DavidDeSimone/ng-js

If you want to make a module that is statically linked to emacs-ng, then you should start by looking in the rust_src directory of emacs-ng (https://github.com/emacs-ng/emacs-ng/tree/master/rust_src). From there, I recommend looking at the content of `crates`. If you look at crates, you will see several examples. Looking further at how lsp_json is organized, you will see a crate with a Cargo.toml and a  buid.rs. If you look at jsp_json's lib.rs, you'll see some boiler plate, and in `parsing.rs` you'll see examples of of functions that are exposed to emacs via our proc macro `#[lisp_fn]`. You'll also see examples of us defining symbols, along with some some code used for exporting functions to the C layer:

```rs
#[allow(dead_code)]
fn init_syms() {
    def_lisp_sym!(QCnull, ":null");
    def_lisp_sym!(QCfalse, ":false");
    def_lisp_sym!(QCobject_type, ":object-type");
    def_lisp_sym!(QCarray_type, ":array-type");
    def_lisp_sym!(QCnull_object, ":null-object");
    def_lisp_sym!(QCfalse_object, ":false-object");
    def_lisp_sym!(QCser_null_object, ":ser-null-object");
    def_lisp_sym!(QCser_false_object, ":ser-false-object");
    def_lisp_sym!(QCjson_config, ":json-config");
    def_lisp_sym!(Qalist, "alist");
    def_lisp_sym!(Qplist, "plist");
    def_lisp_sym!(Qarray, "array");
}

include!(concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/out/parsing_exports.rs"
));
```

Does that answer your question? I can help if you have any follow ups.

**Reply by thiskappaisgrey** [reply]
Yeah! Thank you so much! I'll definitely check those out. I wanted to make rust functions that are dynamically linked, so I'll try checking out `emacs-module-rs`.
